My Site Dashboard: Tabs - Custom view with QS focus point, introduce `TabsUiState`

ashiagr opened the pull request:
Parent #16006

This PR
- Sets custom view for my site tabs to allow showing a QS focus point over `Site Menu` tab in the future.
- Refactors code to update tabs using `TabsUiState`.

/cc @tiagomar (This is mostly a preparatory PR for introducing `Site Menu` QS step with QS focus point in the next PR and can be skipped for testing till that point.)

To test:

Test.1 - Tabs enabled - Tabs UI 

1. Launch the app with a wpcom account having sites.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn on the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app
5. Logout/login and select "Show me around" when the quick start prompt is shown.
6. Notice that tabs are shown on `My Site` screen with UI same as before.

Optional Tests (covered in adjusted unit tests - c2979fdc67b468576c5b4d3a9c69ba0155ed6c7c)

Test.2 - Tabs disabled

1. Launch the app with a wpcom account having sites.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn off the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app
5. Logout/login and select "Show me around" when the quick start prompt is shown.
6. Notice that tabs are not shown on `My Site` screen.


Test 3 - Tabs enabled - No Sites

1. Launch the app with a wpcom account having no sites.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn on the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app.
5. Logout/login.
6. Notice that tabs are not shown on `My Site` screen.

## Regression Notes
1. Potential unintended areas of impact
Tabs don't display as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated tested

3. What automated tests I added (or what prevented me from doing so)
Unit tests were updated to reflect the changes.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Add my site tab custom view with quick start focus point widget

ashiagr committed:
Add stringResId for each MySiteTabType

ashiagr committed:
Convert tabTitles to orderedTabTypes

ashiagr committed:
Introduce TabsUiState

ashiagr committed:
Update tabs using tabs ui state

ashiagr committed:
Adjust tests for show tabs from tabs ui state

ashiagr committed:
Introduce mutable tabs ui state live data

Since state (MySiteUiState) livedata is immutable, results from mutable tabs ui state livedata are merged with it and combined results are mapped to update tab ui state in the UiModel.

It will be useful when we need to update tab ui state to show QS focus point in a future PR.

zwarm commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteViewModel.kt:
```diff
@@ -192,39 +193,30 @@ class MySiteViewModel @Inject constructor(
-        } else {
-            buildNoSiteState()
+    val uiModel: LiveData<UiModel> = merge(_tabsUiState, state) { tabsUiState, mySiteUiState ->
+        with(requireNotNull(mySiteUiState)) {
```
I understand why you did this, but there exists the potential to throw an `IllegalArgumentException` with the "requireNotNull".  And yes, the risk is very low, so I am only calling it out that I noticed it.

ashiagr replied:
Answered in https://github.com/wordpress-mobile/WordPress-Android/pull/16142#issuecomment-1075950562.

zwarm reviewed (approved):
@ashiagr, nice job setting up the tab using the mediator and custom tab view. Splitting this out, especially for the focusPoint was a good call. My future self appreciates the effort, as I will be digging more into this with tab experiment implementation. 🥳 . 

I left one comment re: `RequireNotNull` but it won't hold up the merge. 

I think the Test 3 instructions are off. There is no “show me around” prompt when the user has no sites. I was able to verify that the No Sites views are shown as expected and no tabs.

👏 

Updated: I was going to merge this, but the "Not ready for Merge" label is there and I don't see merge instructions. I will leave this for you then.

tiagomar commented:
Hey @ashiagr ! 👋  Thanks for the explanation. =)

I just wanted to confirm if an issue I faced is known/expected at this point.

Tabs enabled - No Sites

1. Launch the app with a wpcom account having no sites.
2. Go to `My Site` -> `Me` -> `App Settings` -> `Debug settings`.
3. Turn on the `MySiteDashboardTabsFeatureConfig` flag under the `Features in development` section.
4. Restart the app
5. Logout/login
6. Notice that tabs are not shown on `My Site` screen.
7. Add a new site
8. Delete the site
**Observed**: Deleted site info is still displayed as well as the tabs until the app is closed.

ashiagr commented:
👋 @zwarm 

> I think the Test 3 instructions are off. 

Oops, I've updated the test case. Thanks for confirming that functionality behaves as expected. 🙇‍♀️ 

> I left one comment re: RequireNotNull but it won't hold up the merge.
> the "Not ready for Merge" label is there 

Good point @zwarm. I added `RequireNotNull` for `mySiteState`  as `state` livedata is declared with non-null `MySiteUiState` type and the way `merge` function behaves (sending values of the livedata that got updated and the other livedata). Tabs state livedata is not updated at this point which can possibly send a null `state` on merge if it is updated before `state` livedata.

I added `Not ready for Merge` label at the last moment as I wanted to be sure that `RequireNotNull` check doesn't cause any crashes when I start updating the tabs state in the next PR, requiring any change of approach.

I'll take care of the merging, thanks for noticing the label. 🥇 

-------

Thanks for testing, @tiagomar 🙏  
> Delete the site
Observed: Deleted site info is still displayed as well as the tabs until the app is closed.

This is a known issue and is currently being worked upon: https://github.com/wordpress-mobile/WordPress-Android/issues/15592.

ashiagr commented:
I'm merging this PR now after adding appropriate checks and testing `merge` and `requireNotNull` check approach (added in f67fbe3) doesn't cause any crashes when `tabs ui state` livedata is updated in PR https://github.com/wordpress-mobile/WordPress-Android/pull/16159.
